flexprep/domain/processing.py

```diff
@@ -41,57 +42,93 @@ def __init__(self) -> None:
+        if result is None:
+            logging.error("Failed to sort and download files.")
             return
 
-        temp_files = [self.s3_client.download_file(to_process)]
-        prev_file = file_objs[1]
-        prev_step = int(prev_file["step"])
-
-        if prev_step != 0:
-            temp_files.append(self.s3_client.download_file(prev_file))
-            init_files = file_objs[2:4]
-        else:
-            init_files = file_objs[1:3]
+        temp_files, to_process, prev_file = result
 
-        init_files.sort(key=lambda x: x["key"][-2:])
-        temp_files.extend([self.s3_client.download_file(f) for f in init_files])
+        ds_in = self._load_and_validate_data(temp_files, to_process, prev_file)
+        if ds_in is None:
+            logging.error("Failed to load and validate data.")
+            return
+        ds_out = self._apply_flexpart(ds_in)
+        self._save_output(
+            ds_out, to_process["forecast_ref_time"], int(to_process["step"])
+        )
 
+    def _sort_and_download_files(
+        self, file_objs: list[FileObject]
+    ) -> Optional[tuple[list[str], FileObject, FileObject]]:
+        """Sort file objects, validate, and select files for processing."""
+        try:
+            sorted_files = sorted(file_objs, key=lambda x: int(x["step"]), reverse=True)
+            if len(sorted_files) < 3:
+                raise ValueError("Not enough files for pre-processing")
+            to_process = sorted_files[0]
+            prev_file = sorted_files[1]
+            if int(prev_file["step"]) != 0:
+                init_files = sorted_files[2:4]
+                tempfiles = self._download_files(
+                    [to_process, prev_file, init_files[0], init_files[1]]
+                )
+            else:
+                init_files = sorted_files[1:3]
+                tempfiles = self._download_files(
+                    [to_process, init_files[0], init_files[1]]
+                )
+            return tempfiles, to_process, prev_file
+        except Exception as e:
+            logging.error(f"Sorting and validation failed: {e}")
+            return None
+
+    def _download_files(self, files_to_download: list[FileObject]) -> list[str]:
+        """Download files from S3 based on the file objects."""
+        try:
+            temp_files = [
+                self.s3_client.download_file(file_obj) for file_obj in files_to_download
+            ]
+            return temp_files
+        except Exception as e:
+            logging.error(f"File download failed: {e}")
+            return []
+
+    def _load_and_validate_data(
+        self, temp_files: list[str], to_process: FileObject, prev_file: FileObject
+    ) -> typing.Any:
+        """Load and validate data from downloaded files."""
         request = {"param": list(CONSTANTS | INPUT_FIELDS)}
         with config.set_values(data_scope="ifs"):
             source = data_source.DataSource(datafiles=temp_files)
```

Comment:
If the S3 bucket supports pre signed URLs, we could use that to stream the data straight into memory. Note this would require an extension of the data source module but it's fairly small. I can create a ticket to track this.

Reply:
I was able to create a presigned URL, but I couldn't access it due to an "Access Denied" error. I've opened a ticket with EWC to clarify this matter.

Reply:
So, the presigned URLs should work fine on EWC @cfkanesan . I suggest we address this in another PR.

Reply:
Sounds good to me